app/validators/rules/rule_validator.py

```diff
@@ -0,0 +1,207 @@
+            self._validate_self_references(
+                rules, operator_name, allow_self_reference=allow_self_reference
+            )
+
+        self._validate_options(rules, operator_name)
+
+    def _validate_self_references(self, rules, operator_name, *, allow_self_reference):
+        """
+        Validate references to `self` are within the context of the `map` operator
+        """
+        arguments_for_non_map_operators = self._get_flattened_arguments_for_non_map_operators(
+            rules[operator_name]
+        )
+        if SELF in arguments_for_non_map_operators and not allow_self_reference:
+            self.add_error(self.SELF_REFERENCE_OUTSIDE_MAP_OPERATOR, rule=rules)
+
+    def _validate_map_operator(self, operator):
+        """
+        Validates that the first argument to the map operator has a reference to `self`
+        The second argument is currently not validated here as it can currently only be `date-range`
+        """
+        function_to_map_over_arguments = list(operator["map"][0].values())[0]
+        if SELF in function_to_map_over_arguments:
+            return None
+
+        arguments_for_non_map_operators = self._get_flattened_arguments_for_non_map_operators(
+            function_to_map_over_arguments
+        )
+
+        if SELF not in arguments_for_non_map_operators:
+            self.add_error(
+                self.MAP_OPERATOR_WITHOUT_SELF_REFERENCE, rule=operator["map"][0]
+            )
+
+    def _get_flattened_arguments_for_non_map_operators(self, arguments):
+        non_operator_arguments = []
+        for argument in arguments:
+            if isinstance(argument, dict) and any(
+                operator in argument
+                for operator in ALL_OPERATORS
+                if operator != Operator.MAP
+            ):
+                for operands in argument.values():
+                    non_operator_arguments += self._get_flattened_arguments_for_non_map_operators(
+                        operands
+                    )
+            else:
+                non_operator_arguments.append(argument)
+
+        return non_operator_arguments
+
+    def _validate_date_operator(self, operator):
+        """
+        Validates that when an answer value source is used it is a date
+        """
+        first_argument = operator["date"][0]
+        if (
+            isinstance(first_argument, dict)
+            and first_argument.get("source") == "answers"
+            and self.questionnaire_schema.get_answer(first_argument["identifier"])[
+                "type"
+            ]
+            not in ["Date", "MonthYearDate", "YearDate"]
+        ):
+            self.add_error(
+                self.DATE_OPERATOR_REFERENCES_NON_DATE_ANSWER,
+                value_source=first_argument,
+            )
+
+    def _validate_count_operator(self, operator):
+        """
+        Validates that when an answer value source is used it is a checkbox
```

Comment:
The doc string and name of function don't really go together

Reply:
Didn't clock it since it was existing but have made a slight change for clarity.